README.rst
===========================================
Django-ROA (Resource Oriented Architecture)
===========================================

**Use Django's ORM to model remote API resources.**

Fork of original `David Larlet Django ROA lib <http://code.larlet.fr/django-roa/src>`_.
Now ROA works directly with an API like `Django Rest Framework <http://www.django-rest-framework.org/>`_

**How does it works:**
Each time a request is passed to the database, the request is intercepted and transformed to an HTTP request to the remote server with the right
method (GET, POST, PUT or DELETE) given the get_resource_url_* methods specified in the model's definition.


Documentation
=============

Initial documentation:

- `Wiki home <http://code.larlet.fr/django-roa/wiki/Home>`_
- `Getting started with Django-ROA <http://code.larlet.fr/django-roa/wiki/GettingStarted#!getting-started-with-django-roa>`_
- `Developing with Django-ROA <http://code.larlet.fr/django-roa/wiki/Development#!developing-with-django-roa>`_


Supported versions
==================

.. image:: https://travis-ci.org/charles-vdulac/django-roa.png?branch=master
    :target: https://travis-ci.org/charles-vdulac/django-roa
    :alt: Build Status

- Django 1.4, 1.5, 1.6, 1.7
- Python 2.6, 2.7


Installation
============

.. code:: bash

    $ pip install -e git+https://github.com/charles-vdulac/django-roa.git@master#egg=django_roa


Fork getting started
====================

If you have an API output like this (typical DRF output):

.. code:: python

    # GET http://api.example.com/articles/
    # HTTP 200 OK
    # Content-Type: application/json
    # Vary: Accept
    # Allow: GET, POST, HEAD, OPTIONS

    {
        "count": 3,
        "next": null,
        "previous": null,
        "results": [
            {
                "id": 1,
                "headline": "John's first story",
                "pub_date": "2013-01-04",
                "reporter": {
                    "id": 1,
                    "account": {
                        "id": 1,
                        "email": "john@example.com"
                    },
                    "first_name": "John",
                    "last_name": "Smith"
                }
            },
            ...
        ]
    }

Your code will look like this:

.. code:: python

    from django.db import models
    from django_roa import Model as ROAModel

    class Article(ROAModel):
        id = models.IntegerField(primary_key=True)  # don't forget it !
        headline = models.CharField(max_length=100)
        pub_date = models.DateField()
        reporter = models.ForeignKey(Reporter, related_name='articles')

        api_base_name = 'articles'

        @classmethod
        def serializer(cls):
            from .serializers import ArticleSerializer
            return ArticleSerializer

        @classmethod
        def get_resource_url_list(cls):
            return u'http://api.example.com/{base_name}/'.format(
                base_name=cls.api_base_name,
            )

        def get_resource_url_count(self):
            return self.get_resource_url_list()

.. code:: python

    from rest_framework import serializers
    from .models import Article

    class ArticleSerializer(serializers.ModelSerializer):
        reporter = ReporterSerializer()
        class Meta:
            model = Article
            fields = ('id', 'headline', 'pub_date', 'reporter')

Refer to `tests <examples/django_rest_framework/>`_ for full example.

Running tests
=============

- Initial tests: read `documentation <http://code.larlet.fr/django-roa/wiki/GettingStarted#!running-tests>`_
- Fork tests: read `README <examples/django_rest_framework/README.md>`_


Caveats
=======

For the moment, the library doesn't work in this case:

One to one (reversed)
---------------------

.. code:: python

  class Reporter(CommonROAModel):
      account = models.OneToOneField(Account)
      ...

with fixtures:

.. code:: json

    {
        "model": "api.reporter",
        "pk": 1,
        "fields": {
            "first_name": "John",
            "last_name": "Smith",
            "account": 1
        }
    },
    {
        "model": "api.account",
        "pk": 1,
        "fields": {
            "email": "john@example.com"
        }
    },

This works:

.. code:: python

    reporter = Reporter.objects.get(id=1)
    assertEqual(reporter.account.id, 1)
    assertEqual(reporter.account.email, 'john@example.com')

But not this way:

.. code:: python

    account = Account.objects.get(id=1)
    assertEqual(account.reporter.id, 1)
    assertEqual(account.reporter.first_name, "John")


HTTPS certificate pinning
=========================

You can pass ssl args (see `ssl.wrap_socket()`) via the `ROA_SSL_ARGS` of your
``settings.py``.


To pin the server certificate, save the public certificate(s) you want to
pin in *pinned-ca.pem* and add the following to your *settings.py* :

.. code:: python

    from os.path import dirname, join
    ROA_SSL_ARGS = {
        'ca_certs': join(dirname(dirname(__file__)), 'pinned-ca.pem'),
        'cert_reqs': True
    }


CHANGELOG
======================
Django-ROA's changelog
======================

Version 1.8.1, 21 Nov 2014:
--------------------------

* Fix pagination in remote querysets (checking limit_start/limit_stop)
  Thanks to @jathanism


Version 1.8, 12 Sep 2014:
--------------------------

* Django REST Framework support + bug fixes.
  Thanks to @JocelynDelalande, @sandersky, @jathanism and @charles-vdulac


Version 1.7, 11 May 2012:
--------------------------

* @neokeeper aka Martin Skala updated the code in many ways: introduced an
  ROA_FILTERS setting to provide authentication on resources' access,
  improved single instance retrieval, dealt with pks instead of ids when
  possible, updated requirements, code clean up, documentation and
  example updates. Many thanks!
* @NikhilChelliah fixed a bug related to object saving and non-int ids.
  Thanks!


Version 1.6, 16 July 2010:
--------------------------

* Switch to pip's requirements.txt file for external dependencies (restkit and
  piston, note that the later is only useful for tests).
* @cyril fixed a lot of bugs, see changeset 86a68259e83e for details. Thanks!
* @cyrilrbt fixed missing ROA_CUSTOM_ARGS on delete. Thanks!


Version 1.5, 7 January 2010:
----------------------------

* Support for Django 1.2 alpha, no more patch required.
* Handle ManyToMany relations only with the required "through" argument,
  it allows you to define a URL for this resource.
* Update restkit to the latest tip (>0.9.2)
* Update piston to the latest tip (>0.2.3rc1)
* Deprecate "remoteauth" application for now, mapping of M2M relations is not
  fully compatible, you should still be able to use the User model though but
  Permissions and Groups' relation will not be correct anymore.
  This is still a work in progress, suggestions welcome!


Version 1.4, 15 September 2009:
-------------------------------

* Support for named relations.
* Support for proxy models.
* Fix a bug to deal with ModelForms
* Switch from restclient to the latest tip of restkit (>0.8.2)
* Update piston to the latest tip (>0.2.3rc1)
* Introduce a new settings, ROA_HEADERS, to specify headers sent to the server
* Better errors' feedback from the server


Version 1.3, 25 May 2009:
-------------------------

* First Open-Source release! (BSD license)
* Handle ManyToMany relations in a RESTful way.
* Support for unicode.
* Support of custom de/serialization improved.
* Addition of a Django command (inspectresources).
* Add a couple of useful new settings (see documentation).
* Switch tests to django-piston (0.2.1) and to unittests, use of logging.
* Update py-restclient to the latest revision (1.3.2).
* Backward incompatible changes: count is now handled by it's own URL via
  get_resource_url_count (default is get_resource_url_list/count/ but you can
  customize it).


Version 1.2, 30 April 2009:
---------------------------

* Update (Null)Boolean fields, Django doesn't allow anymore null=True.
* Move documentation to BitBucket's wiki.
* Update py-restclient to the latest revision (1.3).


Version 1.1, 28 February 2009:
------------------------------

* Add support for Groups and Permissions in remoteauth application (still
  depends on builtin Django's ContentTypes).
* Add support for ModelForms with complex relations.
* Add an example to document the use of a custom serializer.
* Update py-restclient to the latest revision.
* Backward incompatible changes: RemoteUser model has been renamed to User and
  ROAUserManager to UserManager to be consistent with the existing auth app.


Version 1.0, 23 January 2009:
-----------------------------

* Add support for many-to-many relations.
* Warning: Many-to-many relations depends on Django's issue #10109, apply the
  attached patch if you need it.
* Update py-restclient to 1.1.4.


Version 0.9, 9 January 2009:
----------------------------

* Ease subclassing of MethodDispatcher with custom slugs.


Version 0.8, 7 January 2009:
----------------------------

* Add support for admin options except search_fields because of advanced
  querysets based on Q objects.
* Support for ManyToMany relations in progress.
* Warning: trunk version of Django is required for now, prior to revision 9695
  in order to handle FloatFields
* Bugfixes: ROA_URL_OVERRIDES_* settings are optionnal and ForeignKey fields
  should work as expected.
* Backward incompatible change: admin classes must inherit from
  django_roa.ModelAdmin to be consistent (see documentation and example).


Version 0.7, 2 January 2009:
----------------------------

* Add support for most useful fields (see specifications).
* Warning: there are some Django bugs (hopefully with patches) which are
  required in order to use BooleanFields with None values and JSON
  serialization (#5563) or FloatFields which are considered as unicode
  (#9942). Patch your Django installation if you need those ones, that's why
  actual tests failed.


Version 0.6, 28 December 2008:
------------------------------

* Clean up remoteauth application.
* Update py-restclient to 1.0.1.


Version 0.5, 27 December 2008:
------------------------------

* Backward incompatible change: Remote* classes had been renamed to ROA*, this
  is an internal change which should not affect your code.
* Add support for ROA_URL_OVERRIDES_* settings and a complete example.
* Declaration of default manager is no more required, if you inherit from
  django_roa.Model it will be added automatically given your ROA_MODELS
  setting.
* Use the latest version of restclient (0.2.1), no more httplib2 dependency.
  This version uses pycurl, urllib2 or httplib2 given your configuration, in
  this order of preferences.


Version 0.4, 27 December 2008:
------------------------------

* Backward incompatible change: resource_url_list is no more defined in Meta

django_roa/db/query.py
import logging
from StringIO import StringIO

from django.conf import settings
from django.db.models import query
from django.core import serializers
# Django >= 1.5
from django_roa.db import get_roa_headers

try:
    from django.db.models.constants import LOOKUP_SEP
#Django < 1.4
except:
    from django.db.models.sql.constants import LOOKUP_SEP
from django.db.models.query_utils import Q
from django.utils.encoding import force_unicode

from restkit import Resource, ResourceNotFound
from django_roa.db.exceptions import ROAException, ROANotImplementedYetException

logger = logging.getLogger("django_roa")

ROA_MODEL_NAME_MAPPING = getattr(settings, 'ROA_MODEL_NAME_MAPPING', [])
ROA_ARGS_NAMES_MAPPING = getattr(settings, 'ROA_ARGS_NAMES_MAPPING', {})
ROA_FORMAT = getattr(settings, 'ROA_FORMAT', 'json')
ROA_FILTERS = getattr(settings, 'ROA_FILTERS', {})
ROA_SSL_ARGS = getattr(settings, 'ROA_SSL_ARGS', {})

DEFAULT_CHARSET = getattr(settings, 'DEFAULT_CHARSET', 'utf-8')


class Query(object):
    def __init__(self):
        self.order_by = []
        self.extra_order_by = []
        self.default_ordering = []
        self.filters = {}
        self.excludes = {}
        self.search_term = None
        self.filterable = True
        self.limit_start = None
        self.limit_stop = None
        self.where = False
        self.select_related = False
        self.max_depth = None
        self.extra_select = {}
        self.select_for_update = False

    def can_filter(self):
        return self.filterable

    def clone(self):
        return self

    def clear_ordering(self):
        self.order_by = []

    def filter(self, *args, **kwargs):
        self.filters.update(kwargs)

    def search(self, search_term, limit_start=None, limit_stop=None):
        self.search_term = search_term
        self.limit_start = limit_start
        self.limit_stop = limit_stop

    def exclude(self, *args, **kwargs):
        self.excludes.update(kwargs)

    def set_limits(self, start=None, stop=None):
        self.limit_start = start
        self.limit_stop = stop
        self.filterable = False

    def add_select_related(self, fields):
        """
        Sets up the select_related data structure so that we only select
        certain related models (as opposed to all models, when
        self.select_related=True).
        """
        field_dict = {}
        for field in fields:
            d = field_dict
            for part in field.split(LOOKUP_SEP):
                d = d.setdefault(part, {})
        self.select_related = field_dict
        self.related_select_cols = []
        self.related_select_fields = []

    @property
    def parameters(self):
        """
        Returns useful parameters as a dictionary.
        """
        parameters = {}

        # Filtering
        for k, v in self.filters.iteritems():
            key = '%s%s' % (ROA_ARGS_NAMES_MAPPING.get('FILTER_', 'filter_'), k)
            # v could be an object
            try:
                v = v.id
            except:
                pass

            if key in ROA_ARGS_NAMES_MAPPING:
                parameters[ROA_ARGS_NAMES_MAPPING[key]] = v
            else:
                parameters[key] = v
        for k, v in self.excludes.iteritems():
            key = '%s%s' % (ROA_ARGS_NAMES_MAPPING.get('EXCLUDE_', 'exclude_'), k)
            if key in ROA_ARGS_NAMES_MAPPING:
                parameters[ROA_ARGS_NAMES_MAPPING[key]] = v
            else:
                parameters[key] = v
        if self.search_term:
            parameters[ROA_ARGS_NAMES_MAPPING.get('SEARCH_', 'search')] = self.search_term

        # Ordering
        if self.order_by:
            order_by = ','.join(self.order_by)
            parameters[ROA_ARGS_NAMES_MAPPING.get('ORDER_BY', 'order_by')] = order_by

        # Slicing
        if self.limit_start:
            parameters[ROA_ARGS_NAMES_MAPPING.get('LIMIT_START', 'limit_start')] = self.limit_start
        if self.limit_stop:
            parameters[ROA_ARGS_NAMES_MAPPING.get('LIMIT_STOP', 'limit_stop')] = self.limit_stop

        # Format
        parameters[ROA_ARGS_NAMES_MAPPING.get('FORMAT', 'format')] = ROA_FORMAT

        parameters.update(getattr(settings, 'ROA_CUSTOM_ARGS', {}))
        return parameters

    ##########################################
    # Fake methods required by admin options #
    ##########################################

    def add_fields(self, field_names, allow_m2m=True):
        """ Fake method. """
        pass

    def trim_extra_select(self, names):
        """ Fake method. """
        pass

    def results_iter(self):
        """ Fake method. """
        return []

    def combine(self, rhs, connector):
        """ Fake method. """
        pass

    def has_results(self, *args, **kwargs):
        """ Fake method. """
        return True


class RemoteQuerySet(query.QuerySet):
    """
    QuerySet which access remote resources.
    """
    def __init__(self, model=None, query=None):
        self.model = model
        self.query = query or Query()
        self._result_cache = None
        self._iter = None
        self._sticky_filter = False
        self._db = False
        self._for_write = False
        self._hints = {}

        self.params = {}

        self._prefetch_related_lookups = False

    ########################
    # PYTHON MAGIC METHODS #
    ########################

    def __repr__(self):
        if not self.query.limit_start and not self.query.limit_stop:
            data = list(self[:query.REPR_OUTPUT_SIZE + 1])
            if len(data) > query.REPR_OUTPUT_SIZE:
                data[-1] = "...(remaining elements truncated)..."
        else:
            data = list(self)
        return repr(data)

    ####################################
    # METHODS THAT DO RESOURCE QUERIES #
    ####################################

    def iterator(self):
        """
        An iterator over the results from applying this QuerySet to the
        remote web service.
        """
        resource = Resource(self.model.get_resource_url_list(),
                            filters=ROA_FILTERS, **ROA_SSL_ARGS)
        try:
            parameters = self.query.parameters
            logger.debug(u"""Requesting: "%s" through %s with parameters "%s" """ % (
                          self.model.__name__,
                          resource.uri,
                          force_unicode(parameters)))
            response = resource.get(headers=self._get_http_headers(), **parameters)
        except ResourceNotFound:
            return
        except Exception as e:
            raise ROAException(e)

        response = force_unicode(response.body_string()).encode(DEFAULT_CHARSET)

        # Deserializing objects:
        data = self.model.get_parser().parse(StringIO(response))

        # Check limit_start and limit_stop arguments for pagination and only
        # slice data if they are both numeric and there are results left to go.
        # We only perform this check on lists.
        limit_start = getattr(self.query, 'limit_start', None)
        limit_stop = getattr(self.query, 'limit_stop', None)
        if (isinstance(limit_start, int) and isinstance(limit_stop, int) and
           limit_stop - limit_start < len(data) and limit_stop <= len(data) and
           isinstance(data, list)):
                data = data[limit_start:limit_stop]

        # [] is the case of empty no-paginated result
        if data != []:
            serializer = self.model.get_serializer(data=data)
            if not serializer.is_valid():
                raise ROAException(u'Invalid deserialization for %s model: %s' % (self.model, serializer.errors))

            for obj in serializer.object:
                yield obj

    def count(self):
        """
        Returns the number of records as an integer.

        The result is not cached nor comes from cache, cache must be handled
        by the server.
        """
        clone = self._clone()

        # Instantiation of clone.model is necessary because we can't set
        # a staticmethod for get_resource_url_count and avoid to set it
        # for all model without relying on get_resource_url_list
        instance = clone.model()
        resource = Resource(instance.get_resource_url_count(),
                            filters=ROA_FILTERS, **ROA_SSL_ARGS)
        try:
            parameters = clone.query.parameters
            logger.debug(u"""Counting  : "%s" through %s with parameters "%s" """ % (
                clone.model.__name__,
                resource.uri,
                force_unicode(parameters)))
            response = resource.get(headers=self._get_http_headers(), **parameters)
        except Exception as e:
            raise ROAException(e)

        response = force_unicode(response.body_string()).encode(DEFAULT_CHARSET)
        data = self.model.get_parser().parse(StringIO(response))
        return self.model.count_response(data)

    def _get_from_id_or_pk(self, id=None, pk=None, **kwargs):
        """
        Returns an object given an id or pk, request directly with the
        get_resource_url_detail method without filtering on ids
        (as Django's ORM do).
        """
        clone = self._clone()

        # Instantiation of clone.model is necessary because we can't set
        # a staticmethod for get_resource_url_detail and avoid to set it
        # for all model without relying on get_resource_url_list
        instance = clone.model()
        if pk is None:
            instance.id = id
        else:
            instance.pk = pk
        extra_args = {}
        extra_args.update(kwargs)
        extra_args.update(ROA_SSL_ARGS)
        resource = Resource(instance.get_resource_url_detail(),
                            filters=ROA_FILTERS,
                            **extra_args)
        try:
            parameters = clone.query.parameters
            logger.debug(u"""Retrieving : "%s" through %s with parameters "%s" """ % (
                clone.model.__name__,
                resource.uri,
                force_unicode(parameters)))
            response = resource.get(headers=self._get_http_headers(), **parameters)
        except Exception as e:
            raise ROAException(e)

        response = force_unicode(response.body_string()).encode(DEFAULT_CHARSET)

        for local_name, remote_name in ROA_MODEL_NAME_MAPPING:
            response = response.replace(remote_name, local_name)

        # Deserializing objects:
        data = self.model.get_parser().parse(StringIO(response))
        serializer = self.model.get_serializer(data=data)
        if not serializer.is_valid():
            raise ROAException(u'Invalid deserialization for %s model: %s' % (self.model, serializer.errors))

        return serializer.object

    def get(self, *args, **kwargs):
        """
        Performs the query and returns a single object matching the given
        keyword arguments.
        """
        # special case, get(id=X) directly request the resource URL and do not
        # filter on ids like Django's ORM do.

        # keep the custom attribute name of model for later use
        custom_pk = self.model._meta.pk.attname
        # search PK, ID or custom PK attribute name for exact match and get set
        # of unique matches
        attributes_set = set(attr for attr in ['id__exact', 'pk__exact', '%s__exact' % custom_pk] if attr in kwargs.keys())
        exact_match = list(attributes_set)
        # common way of getting particular object
        if kwargs.keys() == ['id']:
            return self._get_from_id_or_pk(id=kwargs['id'])
        # useful for admin which relies on PKs
        elif kwargs.keys() == ['pk']:
            return self._get_from_id_or_pk(pk=kwargs['pk'])
        # check the case of PK attribute with custom name
        elif kwargs.keys() == [custom_pk]:
            return self._get_from_id_or_pk(pk=kwargs[custom_pk])
        # check if there's an exact match filter
        elif len(exact_match) == 1:
            # use the value of exact match filter to retrieve object by PK
            return self._get_from_id_or_pk(pk=kwargs[exact_match[0]])
        else:
            # filter the request rather than retrieve it through get method
            return super(RemoteQuerySet, self).get(*args, **kwargs)

    def latest(self, field_name=None):
        """
        Returns the latest object, according to the model's 'get_latest_by'
        option or optional given field_name.
        """
        latest_by = field_name or self.model._meta.get_latest_by
        assert bool(latest_by), "latest() requires either a field_name parameter or 'get_latest_by' in the model"

        self.query.order_by.append('-%s' % latest_by)
        return self.iterator().next()

    def delete(self):
        """
        Deletes the records in the current QuerySet.
        """
        assert self.query.can_filter(), \
                "Cannot use 'limit' or 'offset' with delete."

        del_query = self._clone()

        # Disable non-supported fields.
        del_query.query.select_related = False
        del_query.query.clear_ordering()

        for obj in del_query:
            obj.delete()

        # Clear the result cache, in case this QuerySet gets reused.
        self._result_cache = None
    delete.alters_data = True

    ##################################################################
    # PUBLIC METHODS THAT ALTER ATTRIBUTES AND RETURN A NEW QUERYSET #
    ##################################################################

    def filter(self, *args, **kwargs):
        """
        Returns a filtered QuerySet instance.
        """
        if args or kwargs:
            assert self.query.can_filter(), \
                    "Cannot filter a query once a slice has been taken."

        clone = self._clone()
        clone.query.filter(*args, **kwargs)
        return clone

    def exclude(self, *args, **kwargs):
        """
        Returns a filtered QuerySet instance.
        """
        if args or kwargs:
            assert self.query.can_filter(), \
                    "Cannot filter a query once a slice has been taken."

        clone = self._clone()
        clone.query.exclude(*args, **kwargs)
        return clone

    def search(self, search_term, limit_start=None, limit_stop=None, *args, **kwargs):
        """
        Returns a filtered QuerySet instance.
        """
        assert not(args or kwargs), "Search accept only one arg (search_term) and limit_start/limit_stop kwargs"

        clone = self._clone()
        clone.query.search(search_term, limit_start=limit_start, limit_stop=limit_stop)
        return clone

    def complex_filter(self, filter_obj):
        """
        Returns a new QuerySet instance with filter_obj added to the filters.

        filter_obj can be a Q object (or anything with an add_to_query()
        method) or a dictionary of keyword lookup arguments.

        This exists to support framework features such as 'limit_choices_to',
        and usually it will be more natural to use other methods.
        """
        if isinstance(filter_obj, Q) or hasattr(filter_obj, 'add_to_query'):
            raise ROAException('Not implemented yet')
        return self.filter(**filter_obj)

    def select_related(self, *fields, **kwargs):
        """
        Returns a new QuerySet instance that will select related objects.

        If fields are specified, they must be ForeignKey fields and only those
        related objects are included in the selection.
        """
        depth = kwargs.pop('depth', 0)
        if kwargs:
            raise TypeError('Unexpected keyword arguments to select_related: %s'
                    % (kwargs.keys(),))
        obj = self._clone()
        if fields:
            if depth:
                raise TypeError('Cannot pass both "depth" and fields to select_related()')
            obj.query.add_select_related(fields)
        else:
            obj.query.select_related = True
        if depth:
            obj.query.max_depth = depth
        return obj

    def order_by(self, *field_names):
        """
        Returns a QuerySet instance with the ordering changed.
        """
        assert self.query.can_filter(), \
                "Cannot reorder a query once a slice has been taken."

        clone = self._clone()
        for field_name in field_names:
            clone.query.order_by.append(field_name)
        return clone

    def extra(self, select=None, where=None, params=None, tables=None,
              order_by=None, select_params=None):
        """
        Only to handle the case of the "cute trick" used in ModelForms (and
        per extension admin) for unique and date constraints.

        Example: ``.extra(select={'a': 1}).values('a').order_by()``.

        http://code.djangoproject.com/browser/django/trunk/django/forms/models.py#L322
        is an interesting documentation for details.
        """
        assert self.query.can_filter(), \
                "Cannot change a query once a slice has been taken"
        if select == {'a': 1}:
            # Totally hackish but we need a fake object to deal with
            # successive calls to values and order_by based on a count
            # which is the less expensive action for our implementation.
            class FakeInt(object):
                def __init__(self, count):
                    self.count = count

                def values(self, *fields):
                    if fields == ('a',): # double check that it's our case
                        return self

                def order_by(self):
                    return self.count

            return FakeInt(self.count())
        raise ROANotImplementedYetException('extra is not yet fully implemented.')

    ###################
    # PRIVATE METHODS #
    ###################

    def _clone(self, klass=None, setup=False, **kwargs):
        if klass is None:
            klass = self.__class__
        query = self.query.clone()
        if self._sticky_filter:
            query.filter_is_sticky = True
        c = klass(model=self.model, query=query)
        c.__dict__.update(kwargs)
        if setup and hasattr(c, '_setup_query'):
            c._setup_query()
        return c

    def _as_url(self):
        """
        Returns the internal query's URL and parameters

        as (u'url', {'arg_key': 'arg_value'}).
        """
        return self.model.get_resource_url_list(), self.query.parameters


    def _get_http_headers(self):
        return get_roa_headers()
